src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/DocDisplayIdleCommands.java

```diff
@@ -0,0 +1,204 @@
+{
+   public DocDisplayIdleCommands()
+   {
+      PREVIEW_LINK = previewLink();
+      PREVIEW_LATEX = previewLatex();
+   }
+   
+   // Private Methods ----
+   
+   // Latex Preview ----
+   
+   private IdleCommand previewLatex()
+   {
+      return new IdleCommand()
+      {
+         @Override
+         public void execute(DocDisplay docDisplay, IdleState state)
+         {
+            onPreviewLatex(docDisplay, state);
+         }
+      };
+   }
+   
+   private void onPreviewLatex(DocDisplay docDisplay, IdleState state)
+   {
+      Position position = resolvePosition(docDisplay, state);
+      Range range = MathJaxUtil.getLatexRange(docDisplay, position);
+      if (range == null)
+         return;
+      
+      docDisplay.renderLatex(range);
+   }
+   
+   // Link Preview ----
+   
+   private IdleCommand previewLink()
+   {
+      return new IdleCommand()
+      {
+         @Override
+         public void execute(DocDisplay docDisplay, IdleState idleState)
+         {
+            onPreviewLink(docDisplay, idleState);
+         }
+      };
+   }
+   
+   private void onPreviewLink(DocDisplay docDisplay, IdleState idleState)
+   {
+      Position position = resolvePosition(docDisplay, idleState);
+      Token token = docDisplay.getTokenAt(position);
+      if (token == null)
+         return;
+
+      if (!token.hasType("href"))
+         return;
+
+      Range tokenRange = Range.fromPoints(
+            Position.create(position.getRow(), token.getColumn()),
+            Position.create(position.getRow(), token.getColumn() + token.getValue().length()));
+            
+      String href = token.getValue();
+      if (href.endsWith(".png")  ||
+          href.endsWith(".jpg")  ||
+          href.endsWith(".jpeg") ||
+          href.endsWith(".gif")  ||
+          href.endsWith(".svg"))
+      {
+         onPreviewImage(docDisplay, href, position, tokenRange);
+      }
+   }
+   
+   private void onPreviewImage(DocDisplay docDisplay,
+                               String href,
+                               Position position,
+                               Range tokenRange)
+   {
+      // for local images, construct a path that the server URI redirects
+      // will handle properly
+      String srcPath = href;
+      if (!href.startsWith("http"))
+      {
+         srcPath = href.startsWith("~/")
+               ? "files/" + href.substring(2)
+               : "file_show?path=" + StringUtil.encodeURIComponent(href);
+      }
+      
+      // check to see if we already have a popup showing for this image;
+      // if we do then bail early
+      String encoded = StringUtil.encodeURIComponent(srcPath);
+      Element el = Document.get().getElementById(encoded);
+      if (el != null)
+         return;
+      
+      // construct image el, place in popup, and show
+      ImageElement imgEl = Document.get().createImageElement();
+      imgEl.setAttribute("src", srcPath);
+      imgEl.getStyle().setProperty("maxWidth", "100px");
```

Comment:
do we need to defend against very tall images as well as against very wide ones? we might need to do what we do elsewhere, which is wait for the image to load so that we can find which of its edges is the long one (see `ImageElementEx` which has access to the natural dimensions), then restrict the long edge to some maximum. (it may be possible to restrict the long edge in pure CSS, which would be the most convenient, but IIRC it's difficult to do this without skewing the aspect ratio)